sw/host/opentitanlib/src/util/openocd.rs

```diff
@@ -0,0 +1,256 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::borrow::BorrowMut;
+use std::cell::{Cell, RefCell};
+use std::io::Read;
+use std::io::Write;
+use std::net::TcpStream;
+use std::process::{Child, Command, Stdio};
+use std::thread;
+use std::time::Duration;
+
+use anyhow::{bail, Context, Result};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use crate::dif::lc_ctrl::LcCtrlReg;
+use crate::impl_serializable_error;
+use crate::io::jtag::{Jtag, JtagParams, JtagTap};
+
+/// Port to spawn and connect to OpenOCD over.
+const OPENOCD_TCL_PORT: &str = "6666";
+
+/// Represents an OpenOCD server that we can interact with.
+pub struct OpenOcdServer {
+    /// JTAG parameters
+    opts: JtagParams,
+    /// OpenOCD child process
+    openocd_server_process: RefCell<Option<Child>>,
+    /// Stream to the telnet interface of OpenOCD
+    openocd_socket_stream: Cell<Option<TcpStream>>,
+}
+
+/// Errors related to the OpenOCD server.
+#[derive(Error, Debug, Deserialize, Serialize)]
+pub enum OpenOcdServerError {
+    #[error("OpenOCD server is not running")]
+    OpenOcdNotRunning,
+    #[error("Generic error {0}")]
+    Generic(String),
+}
+impl_serializable_error!(OpenOcdServerError);
+
+impl OpenOcdServer {
+    /// Create a new OpenOcdServer with the given JTAG options without starting OpenOCD.
+    pub fn new(opts: &JtagParams) -> Result<OpenOcdServer> {
+        Ok(OpenOcdServer {
+            opts: opts.clone(),
+            openocd_server_process: None.into(),
+            openocd_socket_stream: None.into(),
+        })
+    }
+
+    /// Spawn the OpenOCD server and connect to it.
+    fn start(&self, tap: JtagTap) -> Result<()> {
+        // Check if a server was already started.
+        if let Some(mut child) = self.openocd_server_process.take() {
+            // If it's still running, it won't have an exit status
+            if child.try_wait()?.is_none() {
+                self.openocd_server_process.replace(Some(child));
+                bail!("OpenOCD server already running");
+            }
+        }
+
+        let mut args = Vec::new();
+
+        // Pass the path to the adapter config file if given.
+        if let Some(cfg) = &self.opts.openocd_adapter_config {
+            args.extend(["-f", &cfg]);
+        }
+
+        // Because these commands are command-line args, they *must* end with
+        // semicolons to satisfy the TCL lexer. When a TCL program is read from
+        // a file, newlines are sufficient to separate statements.
+        args.extend(["-c", "adapter speed 200;"]);
```

Comment:
I don't know what the best way to handle the adapeter speed is, so far it has been hardcoded in various files but maybe we could take this opportunity to clean that up. @jwnrt  @timothytrippel @dmcardle any suggestions?